microdata_tools/validation/steps/metadata_reader.py

```diff
@@ -17,15 +17,41 @@ def _days_since_epoch(date_string: str) -> int:
     return (date_obj - epoch).days
 
 
+def _determine_datatype(value):
+    if isinstance(value, str):
+        return "STRING"
+    elif isinstance(value, int):
+        return "LONG"
+
+
+def _validate_datatype_in_codelist_and_sentinels(
+    code_list, list_name, datatype
+):
+    invalid_codes = []
+    for item in code_list:
+        code = item.get("code")
+        list_datatype = _determine_datatype(code)
+        if list_datatype != datatype:
+            invalid_codes.append(code)
+    if invalid_codes:
+        error_message = (
+            f"Specified dataType for measure ({datatype}) does not match the "
+            f"datatype within the {list_name} ({list_datatype}). Codes with "
+            f"mismatching datatype are: {invalid_codes}"
+        )
```

Comment:
error message is generally good try not to mix dataType/datatype just use "data type" in messages to users. Also the amount of wrong codes might be a full list of 10k codes 😱 So only report on the first 5 in the list or similar to avoid spamming the user

(If possible change the datatype variables to data_type as well)